api/sendfild/service/waitmanager/waitmanager.go

```diff
@@ -0,0 +1,152 @@
+package waitmanager
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/textileio/powergate/v2/lotus"
+	"github.com/textileio/textile/v2/api/sendfild/service/store"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+var log = logging.Logger("sendfil-waitmanager")
+
+type WaitManager struct {
+	clientBuilder lotus.ClientBuilder
+	store         *store.Store
+	confidence    uint64
+	waitTimeout   time.Duration
+	waiting       map[primitive.ObjectID]*WaitRunner
+	waitingLck    sync.Mutex
+	ticker        *time.Ticker
+	mainCtx       context.Context
+	mainCtxCancel context.CancelFunc
```

Comment:
Using contexts quite carefully here and passing them down into the wait runners to ensure things are cleaned up properly.